### PR TITLE
Add paper_trail.update_columns

### PR DESCRIPTION
**Pull request by TylerRick**
Add paper_trail.update_columns so you can record a version when using
`update_columns` (which skips all callbacks, including the `after_update`
callback).

### CODE REVIEW

**Comment by jaredbeck** [comment]
Hi Tyler. I don't have time to review this right now, but a few quick thoughts:

- Please don't change linter config
- Can this be done without a method extraction, or was that necessary because of `AbcSize` cop? Generally, if you can put refactoring in a separate PR, it makes the review easier.

Sorry, that's not much of a review, but it's all I have time for right now.

**Comment by TylerRick** [comment]
> Can this be done without a method extraction, or was that necessary because of AbcSize cop? 
No, it was only necessary to allow reuse without duplication. As you wish. I've added back the duplication.

File: spec/models/widget_spec.rb [hunk 1]
```diff
@@ -285,6 +285,20 @@
...
+    it "creates a version record" do
+      widget = Widget.create
+      expect(widget.versions.count).to eq(1)
+      Timecop.travel 1.second.since # because MySQL lacks fractional seconds precision
+      widget.paper_trail.update_columns(name: "Bugle")
+      # widget.update_attributes(name: "Bugle")
+      expect(widget.versions.count).to eq(2)
+      expect(widget.versions.last.event).to(eq("update"))
+      expect(widget.versions.last.changeset[:name]).to eq([nil, "Bugle"])
+      expect(widget.versions.last.created_at.to_i).to eq(Time.now.to_i)
```
Comment (jaredbeck) [inline spec/models/widget_spec.rb]: We can't guarantee that this expectation will pass. The operation could take more than a second, or cross a second-boundary.

File: spec/models/widget_spec.rb [hunk 2]
```diff
@@ -285,6 +285,20 @@
     end
   end
 
+  describe ".paper_trail.update_columns", versioning: true do
+    it "creates a version record" do
+      widget = Widget.create
+      expect(widget.versions.count).to eq(1)
+      Timecop.travel 1.second.since # because MySQL lacks fractional seconds precision
+      widget.paper_trail.update_columns(name: "Bugle")
+      # widget.update_attributes(name: "Bugle")
```
Comment (jaredbeck) [inline spec/models/widget_spec.rb]: Please delete all commented-out code.

File: spec/models/widget_spec.rb [hunk 3]
```diff
@@ -285,6 +285,20 @@
...
+      expect(widget.versions.count).to eq(2)
+      expect(widget.versions.last.event).to(eq("update"))
+      expect(widget.versions.last.changeset[:name]).to eq([nil, "Bugle"])
```
Comment (jaredbeck) [inline spec/models/widget_spec.rb]: Good, detailed test.

File: spec/models/on/empty_array_spec.rb [hunk 4]
```diff
@@ -21,6 +21,15 @@ module On
       end
     end
 
+    describe ".paper_trail.update_columns" do
+      it "creates a version record" do
+        widget = Widget.create
+        assert_equal 1, widget.versions.length
+        widget.paper_trail.update_columns(name: "Bugle")
+        assert_equal 2, widget.versions.length
```
Comment (jaredbeck) [inline spec/models/on/empty_array_spec.rb]: Just FYI, there is a nice way to "expect a change" in RSpec. You don't have to use it, what you have is fine, just thought I'd share.

```
expect {
  widget.paper_trail.update_columns(name: "Bugle")
}.to(change { widget.versions.length }).from(1).to(2)
```

File: lib/paper_trail/record_trail.rb [hunk 5]
```diff
@@ -406,6 +435,23 @@ def touch_with_version(name = nil)
...
     end
 
+    # Like the `update_column` method from `ActiveRecord::Persistence`, but also
+    # creates a version to record those changes.
+    def update_column(name, value)
+      update_columns(name => value)
+    end
+
+    # Like the `update_columns` method from `ActiveRecord::Persistence`, but also
+    # creates a version to record those changes.
```
Comment (jaredbeck) [inline lib/paper_trail/record_trail.rb]: Good comment. Please add:

```
# @api public
```

File: lib/paper_trail/record_trail.rb [hunk 6]
```diff
@@ -406,6 +435,23 @@ def touch_with_version(name = nil)
       @record.save!(validate: false)
     end
 
+    # Like the `update_column` method from `ActiveRecord::Persistence`, but also
+    # creates a version to record those changes.
```
Comment (jaredbeck) [inline lib/paper_trail/record_trail.rb]: Good comment. Please add:

```
# @api public
```

File: lib/paper_trail/record_trail.rb [hunk 7]
```diff
@@ -296,7 +296,36 @@ def data_for_update
...
       if record_object_changes?
-        data[:object_changes] = recordable_object_changes
+        data[:object_changes] = recordable_object_changes(changes)
+      end
+      add_transaction_id_to(data)
+      merge_metadata_into(data)
+    end
+
+    def record_update_columns(changes)
+      if enabled?
```
Comment (jaredbeck) [inline lib/paper_trail/record_trail.rb]: I'd prefer a "guard clause" here.

```
return unless enabled?
```

File: lib/paper_trail/record_trail.rb [hunk 8]
```diff
@@ -296,7 +296,36 @@ def data_for_update
         data[:created_at] = @record.updated_at
       end
       if record_object_changes?
-        data[:object_changes] = recordable_object_changes
+        data[:object_changes] = recordable_object_changes(changes)
+      end
+      add_transaction_id_to(data)
+      merge_metadata_into(data)
+    end
+
```
Comment (jaredbeck) [inline lib/paper_trail/record_trail.rb]: Please comment all methods. This method's comment should include:

```
# @api private
```

File: lib/paper_trail/record_trail.rb [hunk 9]
```diff
@@ -296,7 +296,36 @@ def data_for_update
...
+    def record_update_columns(changes)
+      if enabled?
+        versions_assoc = @record.send(@record.class.versions_association_name)
+        version = versions_assoc.create(data_for_update_columns(changes))
+        if version.errors.any?
+          log_version_errors(version, :update)
```
Comment (jaredbeck) [inline lib/paper_trail/record_trail.rb]: Nice job including details like this error logging, and the `enabled?` check. :+1:

File: lib/paper_trail/record_trail.rb [hunk 10]
```diff
@@ -296,7 +296,36 @@ def data_for_update
...
+
+    # Returns data for record_update_columns
+    # @api private
+    def data_for_update_columns(changes)
+      data = {
+        event: @record.paper_trail_event || "update",
+        object: recordable_object,
+        whodunnit: PaperTrail.whodunnit
+      }
+      data[:created_at] = Time.now
```
Comment (jaredbeck) [inline lib/paper_trail/record_trail.rb]: Why do we have to set created_at? Won't AR do this?

File: lib/paper_trail/record_trail.rb [hunk 11]
```diff
@@ -406,6 +435,23 @@ def touch_with_version(name = nil)
...
+      update_columns(name => value)
+    end
+
+    # Like the `update_columns` method from `ActiveRecord::Persistence`, but also
+    # creates a version to record those changes.
+    def update_columns(attributes)
+      changes = {}
+      attributes.each do |k, v|
+        changes[k] = [@record[k], v]
+      end
```
Comment (jaredbeck) [inline lib/paper_trail/record_trail.rb]: Let's add a comment somewhere here explaining that we are constructing a hash that is compatible with the hash returned by `ActiveModel::Dirty#changes`.

File: spec/models/widget_spec.rb (hunk 1 above)
Comment (TylerRick) [reply spec/models/widget_spec.rb]: True... should I just use a `Timecop.freeze` to keep `Time.now` fixed? (That's what I did.)

Looks like there's at least one other expectation with the same problem:
```
        expect(widget.paper_trail.version_at(Time.now)).to be_nil
```

**Comment by TylerRick** [comment]
Thanks for the detailed review, @jaredbeck. I've pushed up another commit to fix the things you suggested.

**Comment by jaredbeck** [comment]
Merged. Thanks for your contribution, Tyler!

> .. maybe we should follow up with some documentation in the readme also, or else people will not know this feature exists.

Please follow up with a docs PR, so other people can discover this feature. Thanks!

**Comment by Jakanapes** [comment]
This is excellent, we were just looking at how to solve this issue cleanly.

**Comment by thejosk** [comment]
@jaredbeck Can you please provide documentation for this?

**Comment by jaredbeck** [comment]
> @jaredbeck Can you please provide documentation for this?

As I said above, contributions of documentation are welcome, thanks!

**Pull request merged**